heron/spi/src/java/com/twitter/heron/spi/utils/LauncherUtils.java

```diff
@@ -0,0 +1,176 @@
+   * @return true if scheduling successfully
+   */
+  public static boolean onScheduleAsLibrary(
+      Config config,
+      Config runtime,
+      IScheduler scheduler,
+      PackingPlan packing) {
+    boolean ret = false;
+
+    try {
+      scheduler.initialize(config, runtime);
+      ret = scheduler.onSchedule(packing);
+
+      if (ret) {
+        // Set the SchedulerLocation at last step,
+        // since some methods in IScheduler will provide correct values
+        // only after IScheduler.onSchedule is invoked correctly
+        ret = SchedulerUtils.setLibSchedulerLocation(runtime, scheduler, false);
+      } else {
+        LOG.severe("Failed to invoke IScheduler as library");
+      }
+    } finally {
+      scheduler.close();
+    }
+
+    return ret;
+  }
+
+  /**
+   * Returns a packing plan generated by configured packing class
+   */
+  public static PackingPlan createPackingPlan(final Config config, final Config runtime) {
+    return getImpl().createPackingPlan(config, runtime);
+  }
+
+  /**
+   * Creates and initializes scheduler instance
+   *
+   * @return initialized scheduler instances
+   */
+  public static IScheduler getSchedulerInstance(Config config, Config runtime) {
+    return getImpl().getSchedulerInstance(config, runtime);
+  }
+
+  /**
+   * Builds initial runtime config instance using topology information.
+   *
+   * @return initial runtime config instance
+   */
+  public static Config getPrimaryRuntime(TopologyAPI.Topology topology,
+                                         SchedulerStateManagerAdaptor adaptor) {
+    return Config.newBuilder()
+        .put(Keys.topologyId(), topology.getId())
+        .put(Keys.topologyName(), topology.getName())
+        .put(Keys.topologyDefinition(), topology)
+        .put(Keys.schedulerStateManagerAdaptor(), adaptor)
+        .put(Keys.numContainers(), 1 + TopologyUtils.getNumContainers(topology))
+        .build();
+  }
+
+  /**
+   * Creates a config instance with packing plan info added to runtime config
+   */
+  public static Config createConfigWithPackingDetails(Config runtime, PackingPlan packing) {
+    Config ytruntime;
+    ytruntime = Config.newBuilder()
+        .putAll(runtime)
+        .put(Keys.instanceDistribution(), packing.getInstanceDistribution())
+        .put(Keys.componentRamMap(), packing.getComponentRamDistribution())
+        .put(Keys.numContainers(), 1 + packing.getContainers().size())
+        .build();
+    return ytruntime;
+  }
+
+  private static LauncherUtilImpl getImpl() {
+    return instance;
+  }
+
+  /**
+   * For testing only. Could be used to setup a mock
+   */
+  public static void setInstance(LauncherUtilImpl instance) {
+    LauncherUtils.instance = instance;
+  }
+
+  /**
+   * Implementation of utility methods provided by {@link LauncherUtils}. The instance of this
+   * class are primarily used for mocking and testing.
+   */
+  public static class LauncherUtilImpl {
```

Comment:
Why create an inner static class inside another one instead of just making createPackingPlan non-static in the parent class? That would seem more straight forward, and would allow subclassing.